vmcnet/physics/spin.py

```diff
@@ -0,0 +1,130 @@
+"""Spin calculations."""
+from typing import Callable
+
+import jax.numpy as jnp
+
+from vmcnet.utils.distribute import get_mean_over_first_axis_fn
+from vmcnet.utils.typing import Array, P, ModelApply, SLArray
+
+
+def create_spin_square_expectation(
+    local_spin_exchange: ModelApply[P], nelec: Array, nan_safe: bool = True
+) -> Callable[[P, Array], jnp.float32]:
+    """Create a function which estimates the observable <psi | S^2 | psi> / <psi | psi>.
+
+    We assume a wavefunction of spin-1/2 particles. If the wavefunction psi is given by
+
+        psi(X) = antisymmetrize(F(R) xi(s)),
+
+    where R is the real-space coordinates in X and s are the corresponding spins, and
+    xi(s) is the spin configuration function corresponding to the S_z eigenstate with
+    the first nelec[0] electrons being spin-up and the last nelec[1] electrons being
+    spin-down.
+
+    The S^2 operator can be written as (sum_i S_i)^2, which can be decomposed in terms
+    of the total S_z operator and the raising and lowering operators S_{i+} and S_{i-},
+    i.e. S_{i+} = S_{i,x} + iS_{i,y} and S_{j-} = S_{j,x} - iS_{j,y}.
+
+    We have
+
+        S^2 = S_z^2 - S_z + sum_i S_{i+} * S_{i-} + sum_{i != j} S_{i+} * S_{j-}.
+
+    Direct calculation with the first three terms shows that
+
+        <psi | S_z^2 - S_z + sum_i S_{i+} * S_{i-} | psi>
+            = 0.25 (N_up - N_down)^2 + 0.5 (N_up + N_down).
+
+    The remaining term is computed by integrating a local electron exchange term, as
+    detailed in `create_local_spin_exchange`.
+
+    Args:
+        local_spin_exchange (Callable): a function which computes the local spin
+            exchange term, F(R_{1 <-> 1 + nelec[0]}) / F(R). Must have the signature
+            (params, x) -> local exchange term array with shape (x.shape[0],).
+        nelec (Array): an array of size (2,) with nelec[0] being the number of spin-up
+            electrons and nelec[1] being the number of spin-down electrons.
+        nan_safe (bool, optional): flag which controls if jnp.nanmean is used instead of
+            jnp.mean. Can be set to False when debugging if trying to find the source of
+            unexpected nans. Defaults to True.
+
+    Returns:
+        Callable: a function which computes the S^2 expectation for a wavefunction (not
+        necessarily normalized) given the number of spin-up and spin-down electrons and
+        a collection of samples used to estimate the local spin exchange. Has the
+        signature
+        (params, x) -> <psi | S^2 | psi> / <psi | psi>
+    """
+    mean_fn = get_mean_over_first_axis_fn(nan_safe=nan_safe)
+
+    def spin_square_expectation(params: P, x: Array) -> jnp.float32:
+        local_spin_exchange_out = local_spin_exchange(params, x)
```

Comment:
This is a good point, I think it's a bit overcomplicated as-is. I'm thinking that I'm going to leave it though, so that the `create_spin_square_expectation` doesn't need to handle the complexity of different possible local_spin_exchange calculations so that I don't have to add a different flag there when I add the double-sum version of the calculation, and I'm going to go ahead and move the N_up * N_down factor into the local spin exchange integral to compensate.